### ISSUE TITLE
EMQX 5.0.22 recv_proxy_info_error

### ISSUE THREAD

**Issue by dsilletti** [emqx/emqx#10488]
### What happened?

I see many of these messages in the logs:
```
[error] supervisor: 'esockd_connection_sup - <0.2336.0>', errorContext: connection_shutdown, reason: {recv_proxy_info_error,closed}, offender: [{pid,<0.31086.31>},{name,connection},{mfargs,{emqx_connection,start_link,[#{enable_authn => true,limiter => #{bytes_in => #{capacity => 1099511627776,initial => 0,rate => infinity},client => #{bytes_in => #{capacity => 1099511627776,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},connection => #{capacity => 1099511627776,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},message_in => #{capacity => 1099511627776,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},message_routing => #{capacity => 1099511627776,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity}},connection => #{capacity => 1000,initial => 0,rate => 100.0},message_in => #{capacity => 1099511627776,initial => 0,rate => infinity},message_routing => #{capacity => 1099511627776,initial => 0,rate => infinity}},listener => {tcp,default},zone => default}]}}]
```

### What did you expect to happen?

No error message

### How can we reproduce it (as minimally and precisely as possible)?

emqx/emqx:5.0.22 official docker container
haproxy:2.7.6 official docker container

proxy protocol configuration

haproxy config extract:
```
listen mqtt
	bind *:1883
	mode tcp
	balance leastconn
	timeout connect 5s
	timeout client 3h
	timeout server 3h
	server emqx emqx:1883 check send-proxy-v2

listen mqtt-tls
	bind *:8883 ssl crt /usr/local/etc/haproxy/certs/mycert.pem
	mode tcp
	balance leastconn
	timeout connect 5s
	timeout client 3h
	timeout server 3h
	server emqx emqx:1883 check send-proxy-v2
```

emqx listener configuration:

```
listeners {

  tcp {
    default {
      acceptors = 16
      bind = ":1883"
      enable_authn = true
      enabled = true
      limiter {
        bytes_in {
          capacity = "infinity"
          initial = "0"
          rate = "infinity"
        }
        client {
          bytes_in {
            capacity = "infinity"
            divisible = false
            failure_strategy = "force"
            initial = "0"
            low_watermark = "0"
            max_retry_time = "10s"
            rate = "infinity"
          }
          connection {
            capacity = "infinity"
            divisible = false
            failure_strategy = "force"
            initial = "0"
            low_watermark = "0"
            max_retry_time = "10s"
            rate = "infinity"
          }
          message_in {
            capacity = "infinity"
            divisible = false
            failure_strategy = "force"
            initial = "0"
            low_watermark = "0"
            max_retry_time = "10s"
            rate = "infinity"
          }
          message_routing {
            capacity = "infinity"
            divisible = false
            failure_strategy = "force"
            initial = "0"
            low_watermark = "0"
            max_retry_time = "10s"
            rate = "infinity"
          }
        }
        connection {
          capacity = 1000
          initial = "0"
          rate = "1000/s"
        }
        message_in {
          capacity = "infinity"
          initial = "0"
          rate = "infinity"
        }
        message_routing {
          capacity = "infinity"
          initial = "0"
          rate = "infinity"
        }
      }
      max_connections = 1024000
      proxy_protocol = true
      proxy_protocol_timeout = "3s"
      tcp_options {
        active_n = 100
        backlog = 1024
        buffer = "4KB"
        high_watermark = "1MB"
        nodelay = true
        reuseaddr = true
        send_timeout = "15s"
        send_timeout_close = true
      }
      zone = "default"
    }
  }
  ws {
    default {
      acceptors = 16
      bind = ":8083"
      enable_authn = true
      enabled = true
      limiter {
        bytes_in {
          capacity = "infinity"
          initial = "0"
          rate = "infinity"
        }
        client {
          bytes_in {
            capacity = "infinity"
            divisible = false
            failure_strategy = "force"
            initial = "0"
            low_watermark = "0"
            max_retry_time = "10s"
            rate = "infinity"
          }
          connection {
            capacity = "infinity"
            divisible = false
            failure_strategy = "force"
            initial = "0"
            low_watermark = "0"
            max_retry_time = "10s"
            rate = "infinity"
          }
          message_in {
            capacity = "infinity"
            divisible = false
            failure_strategy = "force"
            initial = "0"
            low_watermark = "0"
            max_retry_time = "10s"
            rate = "infinity"
          }
          message_routing {
            capacity = "infinity"
            divisible = false
            failure_strategy = "force"
            initial = "0"
            low_watermark = "0"
            max_retry_time = "10s"
            rate = "infinity"
          }
        }
        connection {
          capacity = 1000
          initial = "0"
          rate = "1000/s"
        }
        message_in {
          capacity = "infinity"
          initial = "0"
          rate = "infinity"
        }
        message_routing {
          capacity = "infinity"
          initial = "0"
          rate = "infinity"
        }
      }
      max_connections = 1024000
      proxy_protocol = true
      proxy_protocol_timeout = "3s"
      tcp_options {
        active_n = 100
        backlog = 1024
        buffer = "4KB"
        high_watermark = "1MB"
        nodelay = true
        reuseaddr = true
        send_timeout = "15s"
        send_timeout_close = true
      }
      websocket {
        allow_origin_absence = true
        check_origin_enable = false
        check_origins = "http://localhost:18083, http://127.0.0.1:18083"
        compress = false
        deflate_opts {
          client_context_takeover = "takeover"
          client_max_window_bits = 15
          mem_level = 8
          server_context_takeover = "takeover"
          server_max_window_bits = 15
          strategy = "default"
        }
        fail_if_no_subprotocol = true
        idle_timeout = "7200s"
        max_frame_size = "infinity"
        mqtt_path = "/mqtt"
        mqtt_piggyback = "multiple"
        proxy_address_header = "x-forwarded-for"
        proxy_port_header = "x-forwarded-port"
        supported_subprotocols = "mqtt, mqtt-v3, mqtt-v3.1.1, mqtt-v5"
      }
      zone = "default"
    }
  }

}
```

### EMQX version

<details>

```console
$ ./bin/emqx_ctl broker
sysdescr  : EMQX
version   : 5.0.22
datetime  : 2023-04-23T13:35:00.874520747+00:00
uptime    : 4 days, 17 hours, 47 minutes, 36 seconds
```

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
Ubuntu 22.04.2 LTS

$ uname -a
Linux s01 5.15.0-69-generic #76-Ubuntu SMP Fri Mar 17 17:19:29 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

```

</details>

### Log files

<details>
[error] supervisor: 'esockd_connection_sup - <0.2336.0>', errorContext: connection_shutdown, reason: {recv_proxy_info_error,closed}, offender: [{pid,<0.31086.31>},{name,connection},{mfargs,{emqx_connection,start_link,[#{enable_authn => true,limiter => #{bytes_in => #{capacity => 1099511627776,initial => 0,rate => infinity},client => #{bytes_in => #{capacity => 1099511627776,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},connection => #{capacity => 1099511627776,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},message_in => #{capacity => 1099511627776,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},message_routing => #{capacity => 1099511627776,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity}},connection => #{capacity => 1000,initial => 0,rate => 100.0},message_in => #{capacity => 1099511627776,initial => 0,rate => infinity},message_routing => #{capacity => 1099511627776,initial => 0,rate => infinity}},listener => {tcp,default},zone => default}]}}]
</details>

**Comment by savonarola** [comment]
Trying to reproduce, I made the following docker compose file:

```
version: '3'

services:
  emqx:
    image: emqx:5.0.22
    container_name: emqx
    environment:
    - "EMQX_NODE_NAME=emqx@node1.emqx.io"
    - "EMQX_LOG__CONSOLE_HANDLER__ENABLE=true"
    - "EMQX_LOG__CONSOLE_HANDLER__LEVEL=warning"
    - "EMQX_LISTENERS__TCP__DEFAULT__PROXY_PROTOCOL=true"
    healthcheck:
      test: ["CMD", "/opt/emqx/bin/emqx_ctl", "status"]
      interval: 5s
      timeout: 25s
      retries: 5
    networks:
      emqx-bridge:
        aliases:
        - node1.emqx.io
    ports:
      - 18083:18083
      - 1883:1883


  haproxy:
    image: haproxy:2.7.6
    container_name: haproxy
    volumes:
      - ./haproxy.cfg:/usr/local/etc/haproxy/haproxy.cfg
      - ./haproxy.crt:/usr/local/etc/haproxy/certs/haproxy.crt
      - ./haproxy.key:/usr/local/etc/haproxy/certs/haproxy.crt.key
    networks:
      emqx-bridge:
        aliases:
        - haproxy.emqx.io
    ports:
      - 1884:1884
      - 8883:8883

networks:
  emqx-bridge:
    driver: bridge
```

haproxy.cfg is
```
listen mqtt
	bind *:1884
	mode tcp
	balance leastconn
	timeout connect 5s
	timeout client 3h
	timeout server 3h
	server emqx emqx:1883 check send-proxy-v2

listen mqtt-tls
	bind *:8883 ssl crt /usr/local/etc/haproxy/certs/haproxy.crt
	mode tcp
	balance leastconn
	timeout connect 5s
	timeout client 3h
	timeout server 3h
	server emqx emqx:1883 check send-proxy-v2
```
haproxy.key and haproxy.crt are generated with the commands:
```bash
openssl genrsa -out haproxy.key 1024
openssl req -new -key haproxy.key -out haproxy.csr
openssl x509 -req -days 365 -in haproxy.csr -signkey haproxy.key -out haproxy.crt
```

I cannot observe the described behaviour when connecting to Haproxy ports (1883, 8883) with MQTT clients.

To reproduce the error exactly, I had to explicitly craft a malformed proxy v2 header and send it to the _EMQX port directly_.
```erlang
-module(tcp_proxy_v2).
-export([send_invalid_header/2]).

-define(PROXY_V2_SIG, <<16#0D,16#0A,16#0D,16#0A,16#00,16#0D,16#0A,16#51,16#55,16#49,16#54,16#0A>>).
-define(INVALID, <<?PROXY_V2_SIG/binary, 2:4, 1:4, 1:4, 1:4, 12:16>>).

send_invalid_header(Host, Port) ->
    {ok, Socket} = gen_tcp:connect(Host, Port, [binary, {active, false}]),
    Packet = <<?INVALID/binary>>,
    ok = gen_tcp:send(Socket, Packet),
    ok = gen_tcp:close(Socket).
```

```erlang
> c(tcp_proxy_v2).
{ok,tcp_proxy_v2}
> tcp_proxy_v2:send_invalid_header("127.0.0.1", 1883).
ok
```

It is doubtful that Haproxy can send an invalid header. Are you sure that EMQX port is not exposed anyhow?

**Comment by dsilletti** [comment]
Hi,

EMQX ports are not exposed, this is my relevant compose file (I am using docker in swarm mode but with 1 node only):

```
  haproxy-mqtt:
    image: haproxy:2.7.6
    deploy:
      mode: replicated
      replicas: 1
    ports:
      - target: 1883
        published: 1883
        mode: host 
      - target: 8883
        published: 8883
        mode: host
      - target: 18083
        published: 18083
        mode: host 

  emqx:
    image: emqx/emqx:5.0.22
    deploy:
      mode: replicated
      replicas: 1
```

Thanks

**Comment by savonarola** [comment]
Thank you.
Does your setup have any actual clients?

I  managed to reproduce the situation with the following code:

```erlang
-module(tcp_test).
-export([connect_and_close/2]).

connect_and_close(Host, Port) ->
    {ok, Socket} = gen_tcp:connect(Host, Port, [binary, {active, false}]),
    timer:sleep(1),
    ok = gen_tcp:close(Socket).
```

```erlang
2> c(tcp_test).
{ok,tcp_test}
3> lists:foreach(fun(_) -> io:format("."), tcp_test:connect_and_close("127.0.0.1", 1884) end, lists:seq(1, 100000)).
```

That is, to connect to Haproxy and then disconnect without sending data; do this many times.
I suppose that there is some kind of Haproxy internal optimization: it does not send the whole proxy header at once and does not continue to send it if the client disconnected.

We probably should not interpret this as an unexpected error.

**Comment by savonarola** [comment]
I made an improvement for esockd. It will be updated in EMQX in one of the nearest releases.

**Comment by dsilletti** [comment]
Hi, The error is still there, I am using emqx 5.1.2 at the moment:

```
emqx.1.z7uu2coftvn5@s01    | 2023-07-25T20:41:07.183828+00:00 [error] supervisor: 'esockd_connection_sup - <0.2634.0>', errorContext: connection_shutdown, reason: {recv_proxy_info_error,closed}, offender: [{pid,<0.21269.26>},{name,connection},{mfargs,{emqx_connection,start_link,[#{enable_authn => true,limiter => #{bytes => #{burst => 0,initial => 0,rate => infinity},client => #{bytes => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},connection => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},message_routing => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},messages => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity}},connection => #{burst => 100.0,initial => 0,rate => 100.0},message_routing => #{burst => 0,initial => 0,rate => infinity},messages => #{burst => 0,initial => 0,rate => infinity}},listener => {tcp,default},zone => default}]}}]

emqx.1.z7uu2coftvn5@s01    | 2023-07-25T20:48:23.607307+00:00 [error] supervisor: 'esockd_connection_sup - <0.2634.0>', errorContext: connection_shutdown, reason: {recv_proxy_info_error,closed}, offender: [{pid,<0.22366.26>},{name,connection},{mfargs,{emqx_connection,start_link,[#{enable_authn => true,limiter => #{bytes => #{burst => 0,initial => 0,rate => infinity},client => #{bytes => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},connection => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},message_routing => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},messages => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity}},connection => #{burst => 100.0,initial => 0,rate => 100.0},message_routing => #{burst => 0,initial => 0,rate => infinity},messages => #{burst => 0,initial => 0,rate => infinity}},listener => {tcp,default},zone => default}]}}]

emqx.1.z7uu2coftvn5@s01    | 2023-07-25T20:57:52.163850+00:00 [error] supervisor: 'esockd_connection_sup - <0.2634.0>', errorContext: connection_shutdown, reason: {recv_proxy_info_error,closed}, offender: [{pid,<0.23811.26>},{name,connection},{mfargs,{emqx_connection,start_link,[#{enable_authn => true,limiter => #{bytes => #{burst => 0,initial => 0,rate => infinity},client => #{bytes => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},connection => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},message_routing => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},messages => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity}},connection => #{burst => 100.0,initial => 0,rate => 100.0},message_routing => #{burst => 0,initial => 0,rate => infinity},messages => #{burst => 0,initial => 0,rate => infinity}},listener => {tcp,default},zone => default}]}}]

emqx.1.z7uu2coftvn5@s01    | 2023-07-25T21:01:00.721916+00:00 [error] supervisor: 'esockd_connection_sup - <0.2634.0>', errorContext: connection_shutdown, reason: {recv_proxy_info_error,closed}, offender: [{pid,<0.24273.26>},{name,connection},{mfargs,{emqx_connection,start_link,[#{enable_authn => true,limiter => #{bytes => #{burst => 0,initial => 0,rate => infinity},client => #{bytes => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},connection => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},message_routing => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},messages => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity}},connection => #{burst => 100.0,initial => 0,rate => 100.0},message_routing => #{burst => 0,initial => 0,rate => infinity},messages => #{burst => 0,initial => 0,rate => infinity}},listener => {tcp,default},zone => default}]}}]

emqx.1.z7uu2coftvn5@s01    | 2023-07-25T21:03:44.868641+00:00 [error] supervisor: 'esockd_connection_sup - <0.2634.0>', errorContext: connection_shutdown, reason: {recv_proxy_info_error,closed}, offender: [{pid,<0.24695.26>},{name,connection},{mfargs,{emqx_connection,start_link,[#{enable_authn => true,limiter => #{bytes => #{burst => 0,initial => 0,rate => infinity},client => #{bytes => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},connection => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},message_routing => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},messages => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity}},connection => #{burst => 100.0,initial => 0,rate => 100.0},message_routing => #{burst => 0,initial => 0,rate => infinity},messages => #{burst => 0,initial => 0,rate => infinity}},listener => {tcp,default},zone => default}]}}]

emqx.1.z7uu2coftvn5@s01    | 2023-07-25T21:06:11.016051+00:00 [error] supervisor: 'esockd_connection_sup - <0.2634.0>', errorContext: connection_shutdown, reason: {recv_proxy_info_error,closed}, offender: [{pid,<0.25071.26>},{name,connection},{mfargs,{emqx_connection,start_link,[#{enable_authn => true,limiter => #{bytes => #{burst => 0,initial => 0,rate => infinity},client => #{bytes => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},connection => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},message_routing => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},messages => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity}},connection => #{burst => 100.0,initial => 0,rate => 100.0},message_routing => #{burst => 0,initial => 0,rate => infinity},messages => #{burst => 0,initial => 0,rate => infinity}},listener => {tcp,default},zone => default}]}}]

emqx.1.z7uu2coftvn5@s01    | 2023-07-25T21:20:17.478001+00:00 [error] supervisor: 'esockd_connection_sup - <0.2634.0>', errorContext: connection_shutdown, reason: {recv_proxy_info_error,closed}, offender: [{pid,<0.27204.26>},{name,connection},{mfargs,{emqx_connection,start_link,[#{enable_authn => true,limiter => #{bytes => #{burst => 0,initial => 0,rate => infinity},client => #{bytes => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},connection => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},message_routing => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity},messages => #{burst => 0,divisible => false,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 10000,rate => infinity}},connection => #{burst => 100.0,initial => 0,rate => 100.0},message_routing => #{burst => 0,initial => 0,rate => infinity},messages => #{burst => 0,initial => 0,rate => infinity}},listener => {tcp,default},zone => default}]}}]
```
